notebooks/build_index_alternative.R

```diff
@@ -0,0 +1,181 @@
+# 1) duplicated dataset entries:
+dat <- dat[names(dat) != "eng_asylum_support_ltla.csv"]
+
+# 2) any datasets do not have columns starting with "z_" ?
+which(!mapply(any, Map(grepl, Map(colnames, dat), pattern = "z_")))
+
+# 3) any datasets without "ltla24_code" column ?
+which(!mapply(any, Map(`==`, Map(colnames, dat), "ltla24_code")))
+
+# one dataset had a column named "ltla_cd" instead if "ltla_code"
+# two other datasets had columns named "lad_code" instead of "ltla_code"
+# otherwise the entries matched with ltla24 codes, so assuming mistakes in naming
+colnames(dat[["eng_care_home_residents_ltla24.csv"]])[2] <- "ltla24_code"
+colnames(dat[["eng_asylum_support_ltla24.csv"]])[2] <- "ltla24_code" # NOTE: this data also had fewer rows
+# colnames(dat[["eng_homeless_hh_ltla.csv"]])[2]    <- "ltla24_code"
+
+# --- Merge datasets ----
+# only select ltla24 codes and z score data from each dataset
+dat <- Map(function(x) x[, grepl("ltla24_|z_", colnames(x))], dat)
+
+# merge all the datasets by ltla24_code column
+dat <- Reduce(function(x, y) merge(x, y, by = "ltla24_code", all.x = TRUE), dat)
+
+# z_asylum_support_perc had fewer rows imputing the missing values with 0
+# note that for z-scores this is equivalent with mean imputation
+dat[is.na(dat)] <- 0
+
+# check if all columns are proper z-scores
+# they should have mean equal to 0 and sd equal to 1
+which(mapply(mean, dat[-1]) > 0.01)
+which(mapply(sd, dat[-1]) > 1.01)
+
+# FIXME: looks like z_english_poor_none_perc is not a proper z-score
+# turning it to z-score
+dat$z_english_poor_none_perc <- as.numeric(scale(dat$z_english_poor_none_perc))
+
+# --- Run PCA ----
+# Remove non numeric columns from dataset and scale again
+dat_clean <- dat |>
+  select(-ltla24_code) |>
+  mutate(across(everything(), ~ (.x - mean(.x, na.rm = TRUE)) / sd(.x, na.rm = TRUE)))
+
+# Run KMO and Bartlett's test to ensure suitability of variables for PCA
+# Can confirm KMO > 0.6, indicating variables overlap and there is partial correlation
+# Can confirm Bartlett's p-value is  <0, can reject null hypothesis that there is no correlation
+kmo_result <- KMO(dat_clean)
+bartlett_result <- cortest.bartlett(dat_clean)
+print(kmo_result)
+print(bartlett_result)
+
+# Visually confirm indicators are correlated
+ggcorrplot(cor(dat_clean))
+
+# Instantiate model using varimax which maximises the variance of squared loadings 
+# within each PC so each indicator loads strongly on one PC and weakly on 
+# others thus making each PC more interpretable and meaningful.
+# nfactors is determined by scree plot / # eigenvalues > 1
+pca <- principal(dat_clean, nfactors = 5, rotate = "varimax")
+
+# Check the eigenvalues and amount of variance explained
+eigenvalues <- pca$values
+print(eigenvalues) # First 5 factors have values >1
+variance_proportion <- eigenvalues / sum(eigenvalues)
+print(variance_proportion)
+
+# Plot scree to identify number of PCs to retain
+# Elbow is around 4-5 PCs, 5th PC's eigenvalue is >1 so, retain 5 PCs
```

Comment:
If you keep 4 PC's, which variables map to each component, does it look like this could also make sense?

Reply:
@jennajt this is for you.